src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -15,9 +14,18 @@ namespace System
     // IList<U> and IReadOnlyList<U>, where T : U dynamically.  See the SZArrayHelper class for details.
     public abstract partial class Array : ICloneable, IList, IStructuralComparable, IStructuralEquatable
     {
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern unsafe Array InternalCreate(RuntimeType elementType, int rank, int* pLengths, int* pLowerBounds);
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Array_CreateInstance")]
+        private static unsafe partial void InternalCreate(QCallTypeHandle type, int rank, int* pLengths, int* pLowerBounds, ObjectHandleOnStack retArray);
+
+        private static unsafe Array InternalCreate(RuntimeType type, int rank, int* pLengths, int* pLowerBounds)
+        {
+            Array? retArray = null;
+            InternalCreate(new QCallTypeHandle(ref type), rank, pLengths, pLowerBounds, ObjectHandleOnStack.Create(ref retArray));
+            return retArray!;
+        }
 
+        private static unsafe Array InternalCreateFromArrayType(RuntimeType type, int rank, int* pLengths, int* pLowerBounds)
+            => InternalCreate(type, -rank, pLengths, pLowerBounds);
```

Comment:
subjective nit: using `-` to change the meaning of another argument is used here, but commented in the native code. I wonder whether we could add a short comment here as well (or have two type arguments: `elementType` and `arrayType` and just check for nulls)

Reply:
I am replacing this trick with a self-describing `bool` argument.